Helper setup instructions (#663)
private-attribution/ipa: in-market-test/helper-nodes/setup_instructions.md

Here are the instructions for setting up node for running IPA.

## Step 1: Download code and run a sample test
1. If you do not have Rust installed on your machine, you can follow [these](https://doc.rust-lang.org/book/ch01-01-installation.html#installation) instructions to do so.
2. Download the code from https://github.com/private-attribution/ipa
   You can do this by running the command on the terminal:
   `git clone https://github.com/private-attribution/ipa`
3. This will create a folder named “ipa”. Go inside the folder.

   Command: `cd ipa`
4. Build the Rust code

   Command: `cargo build`
5. Running a sample test :

   Command: `cargo bench --bench oneshot_ipa --features="enable-benches" --no-default-features`

## Step 2: Setup helper configuration
1. Provide details of how the machine can be accessed from public internet. For this, we would need to know
   i.  IP address/DNS which is accessible from public internet (i.e. firewall allows connecting to the helper port (443))
   ii. Confirm if machine is able to listen on port 443

> 2. Download docker image of IPA from shared folder. It we will be uploading an executable docker image which will have IPA code.
>    Link: TBD.  Alternatively, you can build the docker image directly from the Rust code with this command

Reviewer: ```suggestion
2. Download docker image of IPA from shared folder. We will be uploading an executable docker image which will have IPA code.
   Link: TBD.  
   
   Alternatively, you can build the docker image directly from the Rust code with this command
```

   `scripts/helper-image.sh --identity 1 --hostname localhost 1`

3. Docker image already contains (default) public and private keys for encrypting match keys and also,
   to secure the report files while in transit.

   1. Installing/Upgrading Docker
       - If you need to install docker, follow the instructions [here](https://docs.docker.com/engine/install/).
       - If you already have docker, ensure that you are running at least Docker Version 20.10.22.
          You can check version by running command `docker -v` on the terminal.

   2. Create a docker image of the helper

       `cd <path_to_ipa_source_code>`

>        `scripts/helper-image.sh --identity 1 --hostname localhost`

Reviewer: Shouldnt we move this to point 2 with docker image instructions

   3.  Check which docker images are available to run and what is its TAG, run following command:

       `docker images`

       You should be able to see "private-attribution/ipa" in the list along with its TAG

   3. To run the docker container with a bash terminal and see what is inside the image run:

       `docker run -it docker.io/private-attribution/ipa:<IMAGE_TAG> /bin/bash`

>    4. To copy a TLS public key from the docker image to the host :     Open a separate terminal and run following command to see available docker images

Reviewer: ```suggestion
   4. To copy a TLS public key from the docker image to the host :     
   
   Open a separate terminal and run following command to see available docker images
```

       `docker container ls`

       Copy the CONTAINER ID corresponding to your IMAGE.

       You will know what is your Helper party number i.e. 1,2 or 3. Here we show for Helper `1`:

>       `docker cp <CONTAINER_ID>:/etc/ipa/pub/1.pem .  `

Reviewer: Just realized, we now have encryption keys also available. Could you add a line here for copying public key for encryption also?

3. This will create a folder named “ipa”. Go inside the folder.

   Command: `cd ipa`
4. Build the Rust code

   Command: `cargo build`
5. Running a sample test :

   Command: `cargo bench --bench oneshot_ipa --features="enable-benches" --no-default-features`

## Step 2: Setup helper configuration
1. Provide details of how the machine can be accessed from public internet. For this, we would need to know
   i.  IP address/DNS which is accessible from public internet (i.e. firewall allows connecting to the helper port (443))
   ii. Confirm if machine is able to listen on port 443

2. Installing/Upgrading Docker
   - If you need to install docker, follow the instructions [here](https://docs.docker.com/engine/install/).
   - If you already have docker, ensure that you are running at least Docker Version 20.10.22. You can check version by running command `docker -v` on the terminal.

3. Download docker image of IPA from shared folder. It will be uploaded in an executable docker image which will have IPA code.

   Link: TBD.

   Alternatively, you can build the docker image directly from the Rust code with this command

       `cd <path_to_ipa_source_code>`

>        `scripts/helper-image.sh --identity 1 --hostname localhost`

Reviewer 2: this command generates an image for H1, if H2 and H3 want to do it, they need to either use `--identity 2|3` argument or regenerate the keys.

also `localhost` hostname is a setup suitable for local execution, for real helper parties, we need the DNS name here

   i.  IP address/DNS which is accessible from public internet (i.e. firewall allows connecting to the helper port (443))
   ii. Confirm if machine is able to listen on port 443 

> 2. Download docker image of IPA from shared folder. It we will be uploading an executable docker image which will have IPA code. 

Reviewer: Remove "It"

   
   Link: TBD

3. Docker image already contains (default) public and private keys for encrypting match keys and also, 
   to secure the report files while in transit.
  

>    1. If you need to install docker, follow the instructions [here](https://docs.docker.com/engine/install/). 

Reviewer: I still cant follow these instructions. Where do we get the IPA docker image from?

Author: It can be build from the source with this command:  `scripts/helper-image.sh --identity 1 --hostname localhost 1`  

Let me add to the instructions.

>    1. If you need to install docker, follow the instructions [here](https://docs.docker.com/engine/install/). 

Reviewer: ```suggestion
   1. If you need to install docker, follow the instructions [here](https://docs.docker.com/engine/install/). 
   2. If you already have docker, ensure that you are running at least Docker Version 20.10.22. 
       You can check version by running command `docker -v` on the terminal.
```

>    2.  To see what docker images are available run and what their TAGs and IDs are run:

Reviewer: ```suggestion
   2. Create a docker image of the helper 
       cd <path_to_ipa_source_code>
       `scripts/helper-image.sh --identity 1 --hostname localhost`

   3.  Check which docker images are available to run and what is its TAG, run following command:
```

>        `docker images`

Reviewer: ```suggestion
       `docker images`
       You should be able to see "private-attribution/ipa" in the list along with its TAG
```

>    3. To run the docker container with a bash terminal and see what is inside the image run: 
>    
>        `docker run -it docker.io/private-attribution/ipa:<IMAGE_TAG> /bin/bash`

Reviewer: ```suggestion
   3. Run the docker container with a bash terminal: 
   
       `docker run -it docker.io/private-attribution/ipa:<TAG> /bin/bash`
       
       You can also check the content of docker image by running commands like ls, cd etc inside the running docker
       
```

>    4. To copy a TLS public key from the docker image to the host run the following command in a terminal open in the host directory 
>       to which you want to copy the key.  You will know what your Helper party number is yours `{1,2,3}`, here we show for Helper `1`: 
>    
>       `docker cp <IMAGE_ID>:/etc/ipa/pub/1.pem .  `

Reviewer: ```suggestion
   4. To copy a TLS public key from the docker image to the host : 
        Open a separate terminal and run following command to see available docker images
       `docker container ls` 
       Copy the CONTAINER ID corresponding to your IMAGE.
       
       You will know what is your Helper party number i.e. 1,2 or 3. Here we show for Helper `1`:
   
      `docker cp <CONTAINER_ID>:/etc/ipa/pub/1.pem .  `
```

>    5. To copy a TLS public key from Helper `2` from the host onto the docker image run from the directory with the public key `2.pem`: 

Reviewer: I am not sure how to get 2.pem as this wasnt generated in the docker image I made.
